statesync/snapshots.go

```diff
@@ -0,0 +1,260 @@
+type snapshot struct {
+	Height      uint64
+	Format      uint32
+	ChunkHashes [][]byte
+	Metadata    []byte
+
+	trustedAppHash []byte // populated by light client
+}
+
+// Hash generates a snapshot hash, used for lookups. It takes into account not only the height
+// and format, but also the chunk hashes and metadata in case peers have generated snapshots in a
+// non-deterministic manner such that chunks from different peers wouldn't fit together.
+func (s *snapshot) Hash() snapshotHash {
+	// Hash.Write() never returns an error.
+	hasher := sha256.New()
+	hasher.Write([]byte(fmt.Sprintf("%v:%v", s.Height, s.Format)))
+	if s.Metadata != nil {
+		hasher.Write(s.Metadata)
+	}
+	for _, chunkHash := range s.ChunkHashes {
+		hasher.Write(chunkHash)
+	}
+	var hash snapshotHash
+	copy(hash[:], hasher.Sum(nil))
+	return hash
+}
+
+// snapshotPool discovers and aggregates snapshots across peers.
+type snapshotPool struct {
+	stateSource StateSource
+
+	sync.Mutex
+	snapshots     map[snapshotHash]*snapshot
+	snapshotPeers map[snapshotHash]map[p2p.ID]p2p.Peer
+
+	// indexes for fast searches
+	formatIndex map[uint32]map[snapshotHash]bool
+	heightIndex map[uint64]map[snapshotHash]bool
+	peerIndex   map[p2p.ID]map[snapshotHash]bool
+
+	// blacklists for rejected items
+	formatBlacklist   map[uint32]bool
+	heightBlacklist   map[uint64]bool
+	snapshotBlacklist map[snapshotHash]bool
+}
+
+// newSnapshotPool creates a new snapshot pool.
+func newSnapshotPool(stateSource StateSource) *snapshotPool {
+	return &snapshotPool{
+		stateSource:       stateSource,
+		snapshots:         make(map[snapshotHash]*snapshot),
+		snapshotPeers:     make(map[snapshotHash]map[p2p.ID]p2p.Peer),
+		formatIndex:       make(map[uint32]map[snapshotHash]bool),
+		heightIndex:       make(map[uint64]map[snapshotHash]bool),
+		peerIndex:         make(map[p2p.ID]map[snapshotHash]bool),
+		formatBlacklist:   make(map[uint32]bool),
+		heightBlacklist:   make(map[uint64]bool),
+		snapshotBlacklist: make(map[snapshotHash]bool),
+	}
+}
+
+// Add adds a snapshot to the pool, unless the peer has already sent recentSnapshots snapshots. It
+// returns true if this was a new, non-blacklisted snapshot. The snapshot height is verified using
+// the light client, and the expected app hash is set for the snapshot.
+func (p *snapshotPool) Add(peer p2p.Peer, snapshot *snapshot) (bool, error) {
+	appHash, err := p.stateSource.AppHash(snapshot.Height)
+	if err != nil {
+		return false, err
+	}
+	snapshot.trustedAppHash = appHash
+	hash := snapshot.Hash()
+
+	p.Lock()
+	defer p.Unlock()
+
+	switch {
+	case p.formatBlacklist[snapshot.Format]:
+		return false, nil
+	case p.heightBlacklist[snapshot.Height]:
+		return false, nil
+	case p.snapshotBlacklist[hash]:
+		return false, nil
+	case p.snapshots[hash] != nil:
+		p.snapshotPeers[hash][peer.ID()] = peer
+		return false, nil
+	case len(p.peerIndex[peer.ID()]) >= recentSnapshots:
+		return false, nil
+	}
+
+	p.snapshots[hash] = snapshot
```

Comment:
can this override existing snapshot?

Reply:
No, we handle that in the `switch` above:

```go
	case p.snapshots[hash] != nil:
		p.snapshotPeers[hash][peer.ID()] = peer
		return false, nil
```